### ISSUE TITLE
struct_get_names() becomes exponentially slower as struct size increases.

### ISSUE THREAD

**Issue by BscotchSeth** [YoYoGames/GameMaker-Bugs#5058]
### Description

Problem: `struct_get_names()` is slow, and it gets exponentially slower as the struct gets larger.

I have included a test project that shows two ways to get the keys from a struct, along with a timer that records the speed of these operations.

Here is how long it takes to get the names from a struct at different sizes when using `struct_get_names()`.

* 10 variables: .01ms
* 100 variables: .04ms
* 1000 variables: .55ms
* 10000 variables: 26ms
* 100000 variables: 17,420ms (> 17 seconds)

If instead I pre-create an array that is the size of the struct, and then use `struct_foreach()` to iterate over the struct and populate the array with the variable names, this is how long it takes:

* 10 variables: .01ms (about the same amount of time)
* 100 variables: .03ms (25% less time)
* 1000 variables: .19ms (65% less time)
* 10000 variables: 2.17ms (92% less time)
* 100000 variables: 23.8ms (99.9% less time)

Here is the project that shows this:
[StructSpeeds.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/14628795/StructSpeeds.zip)

### Expected Change

The time to call `struct_get_names()` should grow linearly with struct size.

### Steps To Reproduce

1. Start GameMaker
2. Run the project
3. See the issue in the debug log

### How reliably can you recreate this issue using your steps above?

Always

### Which version of GameMaker are you reporting this issue for?

2024.400 (Betas)

### Which platform(s) are you seeing the problem on?

Windows

### Contact Us Package Attached?

- [X] I have attached my Contact Us Package

### Sample Project Added?

- [X] I have included a small sample project

**Comment by BscotchSeth** [comment]
Addendum: This is also true of `ds_map_keys_to_array()` -- it gets exponentially slower as map size increases. In a test I ran, it goes:

* 10 entries: 0.01ms
* 100 entries: 0.03ms
* 1,000 entries: 0.39ms (13x longer than 100 entries)
* 10,000 entries: 25ms (64x longer than 1,000 entries)
* 100,000 entries: 14,630ms (585x longer than 10,000 entries)

**Comment by gnysek** [comment]
Sounds like array which is returned is resized with every entry, instead of being set to total size, but I might be totally wrong.

**Comment by Alphish** [comment]
A little note: in some places you have `struct_foreach` where it seems `struct_get_names` is what you meant (since going by your tests, struct_foreach is what has linear execution time per entry).

In particular:
- at the beginning: "Problem: `struct_foreach()` is slow, and it gets exponentially slower as the struct gets larger."
- in Expected Change: "The time to call struct_foreach() should grow linearly with struct size."

**Comment by BscotchSeth** [comment]
Updated... Thank you!

**Comment by rwkay** [comment]
Fixed in 2024.4 - after changes on my system I see timings of

```
---Testing with 10 entries---
Timer struct_get_names:0.02ms
Timer struct_foreach:0.01ms
---Testing with 100 entries---
Timer struct_get_names:0.01ms
Timer struct_foreach:0.02ms
---Testing with 1000 entries---
Timer struct_get_names:0.06ms
Timer struct_foreach:0.22ms
---Testing with 10000 entries---
Timer struct_get_names:0.57ms
Timer struct_foreach:2.14ms
---Testing with 100000 entries---
Timer struct_get_names:7.55ms
Timer struct_foreach:23.95ms
```

**Comment by BscotchSeth** [comment]
@rwkay That is an insanely huge improvement! Did you also get a chance to look at `ds_map_keys_to_array()`?
